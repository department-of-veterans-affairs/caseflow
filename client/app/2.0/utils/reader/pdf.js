// External Dependencies
import { memoize } from 'lodash';

// Local Dependencies
import { getQueryParams } from 'app/util/QueryParamsUtil';
import { CATEGORIES } from 'store/constants/reader';
import { setCategoryFilter } from 'store/reader/documentList';

/**
 * Method to change to Single Document Mode
 */
export const singleDocumentMode = memoize(() =>
  window.analyticsEvent(CATEGORIES.VIEW_DOCUMENT_PAGE, 'single-document-mode'));

/**
 * Method to navigate between PDF files
 * @param {Object} history -- The Browser Router History Object
 * @param {number} vacolsId -- The ID of the Appeal in VACOLS
 */
export const showPdf = (history, vacolsId) => (documents, docId) =>
  documents[docId] && history.push(`/${vacolsId}/documents/${docId}`);

/**
 * Method to initialize the category Filters
 * @param {array} categories -- The list of categories available
 * @param {string} search -- The Query Search from the URL bar
 * @param {function} dispatch -- The Redux Dispatcher
 */
export const initCategoryFilter = (categories, search, dispatch) => {
  // Calculate the Query Params
  const queryParams = getQueryParams(search);
  const category = queryParams.category;

  // If the category is available apply the filter to the store
  if (categories[category]) {
    dispatch(setCategoryFilter(category, true));

    // Clear out the URI query string params after we determine the initial
    // category filter so that we do not continue to attempt to set the
    // category filter every time routedPdfListView renders.
    return '';
  }

  // Default to return the search so we don't clear
  return search;
};

/**
 * Helper Method to return the Annotation ID
 * @param {Object} annotation -- The ID and TempID of the annotation
 * @returns {string} -- The ID of the annotation
 */
export const keyOfAnnotation = ({ tempId, id }) => tempId || id;

/**
 * Helper Method to calculate the next page index
 * @param {number} pageIndex -- The index of the page
 */
export const pageNumberOfPageIndex = (pageIndex) => pageIndex + 1;

/**
 * Helper Method to calculate the next page number
 * @param {number} pageNumber -- The number of the page
 */
export const pageIndexOfPageNumber = (pageNumber) => pageNumber - 1;

/**
 * Helper Method to determine a valid whole number
 * @param {number} number -- The number to check against
 * @returns {boolean} -- Whether the number is a valid whole number or not
 */
export const isValidWholeNumber = (number) => {
  return !isNaN(number) && number % 1 === 0;
};
