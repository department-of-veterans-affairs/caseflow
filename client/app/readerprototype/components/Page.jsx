import PropTypes from 'prop-types';
import React, { useEffect, useRef } from 'react';

const Page = ({ page, rotation = '0deg', renderItem, scale, storeMetric }) => {
  const canvasRef = useRef(null);
  const scaleFraction = scale / 100;

  const viewport = page.getViewport({ scale: scaleFraction });
  const scaledHeight = viewport.height;
  const scaledWidth = viewport.width;
  let rotatedHeight = scaledHeight;
  let rotatedWidth = scaledWidth;
  let top = 0;
  let left = 0;
  const offset = (scaledHeight - scaledWidth) / 2;

  if (rotation.includes('90') || rotation.includes('270')) {
    rotatedHeight = scaledWidth;
    rotatedWidth = scaledHeight;
    top = `${offset * -1}px`;
    left = `${offset}px`;
  }
  const wrapperStyle = {
    height: `${rotatedHeight}px`,
    width: `${rotatedWidth}px`,
    position: 'relative',
  };
  const canvasStyle = {
    rotate: rotation,
    position: 'relative',
    top,
  };

  useEffect(() => {
    if (canvasRef.current) {
      const renderResult = page
        .render({ canvasContext: canvasRef.current?.getContext('2d'), viewport })
        .promise.catch(() => {
          // this catch is necessary to prevent the error: Cannot use the same canvas during multiple render operations
        });

      renderResult.finally(() => {
        storeMetric(page.pageNumber, page.stats.times);
      });
    }
  }, [canvasRef.current, viewport]);

  return (
    <div id={`canvasWrapper-${page.pageNumber}`} className="prototype-canvas-wrapper" style={wrapperStyle}>
      <canvas
        id={`canvas-${page.pageNumber}`}
        className="prototype-canvas"
        style={canvasStyle}
        ref={canvasRef}
        height={scaledHeight}
        width={scaledWidth}
      />
      {renderItem &&
        renderItem({
          pageNumber: page.pageNumber,
          dimensions: {
            height: scaledHeight,
            width: scaledWidth,
            offsetX: left,
            offsetY: top,
          },
          rotation,
        })}
    </div>
  );
};

Page.propTypes = {
  page: PropTypes.object,
  rotation: PropTypes.string,
  renderItem: PropTypes.func,
  scale: PropTypes.number,
  storeMetric: PropTypes.func,
};

export default Page;
