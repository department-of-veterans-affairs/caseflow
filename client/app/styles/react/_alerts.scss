
.usa-alert {

  &.fixed {
    position: fixed;
    top: 0;
    margin-top: 0;
    z-index: 999;
    left: 50%;
    transform: translate(-50%, 0);
  }

  // Remove after USWDS V.2.0
  &.usa-alert-slim {
    background-size: 4rem;

    .usa-alert-body {
      padding-left: 3.5rem;
    }

    .usa-alert-text {
      font-size: 2rem;
    }
  }

  .usa-alert-text {
    line-height: 1.5;
  }
}
