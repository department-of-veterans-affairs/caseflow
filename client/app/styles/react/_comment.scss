.comment-control-button-container {
  display: flex;
  width: 100%;
  padding: 3px 0;
  justify-content: space-between;
}

.comment-save-button-container {
  display: inline-block;
  width: 100%;
  margin-top: 5px;
  margin-bottom: 17px;
}

.comment-control-button {
  padding: 0;

  &.horizontal {
    text-align: left;
  }
}

.comment-control-button-divider {
  padding: 0 10px;
}

.comment-container-selected {
  box-shadow: 3px 3px 5px $color-gray-light;
}

.comment-container {
  word-wrap: break-word;
  word-break: break-word;
  background-color: $color-white;
  padding: 15px;
  cursor: pointer;

  &:not(.comment-horizontal-container) {
    margin-bottom: 15px;
    border-radius: 3px;
    border: 1px solid $color-gray-lighter;
  }

  &.comment-horizontal-container {
    padding: 25px 25px 25px 40px;
  }
}

// Max width is enforced so the size of this comment box
// fits with existing textareas in style guide
.comment-size-container {
  max-width: 430px;

  .cf-form-textarea {
    margin-bottom: 20px;
  }
}

.comment-horizontal-container {
  display: inline-block;
  width: 100%;
}

.comment-horizontal-spacing {
  margin-right: 49px;
}

.commentIcon-container {
  position: absolute;
  z-index: 10;
  // These dimensions need to match the comment icon SVG dimensions.
  width: 40px;
  height: 40px;
}

