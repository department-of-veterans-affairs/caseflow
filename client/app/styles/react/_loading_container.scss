.loadingContainer-positioning {
  position: absolute;
  width: 100%;
  height: 100%;
}

.loadingContainer-table {
  width: 100%;
  height: 100%;
  display: table;
}

.loadingContainer-table-cell {
  text-align: center;
  display: table-cell;
  vertical-align: middle;
}

.loadingContainer-content {
  position: relative;
  z-index: 1;

  iframe {
    width: 100%;
    height: 80vh;

    &.form9Viewer {
      height: 500px;
    }
  }
}

.loadingContainer-container {
  position: relative;
  overflow: hidden;
}

.cf-loading-button-text {
  display: inline-block;
  margin-right: 10px;
  line-height: 0;
  position: relative;
}

.cf-loading-message {
  .loading-message-text {
    display: inline-block;
    font-style: italic;
    font-weight: bolder;
  }

}

.cf-loading-button-symbol {
  display: inline-block;
}
