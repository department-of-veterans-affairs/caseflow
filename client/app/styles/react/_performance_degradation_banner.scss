.usa-banner {
  background-color: $color-gold-lightest;
  text-align: center;
}

.usa-banner-inner {
  padding: .8em;
}

.banner-icon {
  display: inline-block;
  vertical-align: text-top;
  // height: 25px;
}

.banner-text {
  padding-left: 6px;
}
