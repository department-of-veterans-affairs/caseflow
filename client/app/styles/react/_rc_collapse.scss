@mixin common() {

  width: 0;
  height: 0;
  font-size: 0;
  line-height: 0;
}

@mixin right($w, $h, $color) {

  border-top: $w solid transparent;
  border-bottom: $w solid transparent;
  border-left: $h solid $color;
}

@mixin bottom($w, $h, $color) {

  border-left: $w solid transparent;
  border-right: $w solid transparent;
  border-top: $h solid $color;
}

// scss-lint:disable IdSelector
#arrow {
  @include common();
  @include right(3px, 4px, $color-gray);
  @include bottom(3px, 4px, $color-gray);
}

.rc-collapse {

  &-anim-active {
    transition: height .2s ease-out;
  }

  & > &-item {
    &:first-child {
      border-top: none;
    }
  }

  & > &-item-disabled > .rc-collapse-header {
    cursor: not-allowed;
    color: $color-gray-light;
    background-color: $color-gray-lighter;
  }

  &-content {
    overflow: hidden;

    &-inactive {
      display: none;
    }
  }

  &-item:last-child {
    > .rc-collapse-content {
      border-radius: 0 0 3px 3px;
    }
  }

  & > &-item-active {
    > .rc-collapse-header {

      .arrow {

        #arrow > {
          @include bottom(3px, 4px, $color-gray);
        }
      }
    }
  }
}

.usa-accordion-content {
  &:not([aria-hidden]) {
    position: relative;
    left: 0;
  }
}
