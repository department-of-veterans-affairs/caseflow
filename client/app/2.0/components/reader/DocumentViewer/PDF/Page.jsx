// External Dependencies
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

// Internal Dependencies
import { Comments } from 'components/reader/DocumentViewer/Comments';
import { pageNumber } from 'utils/reader';
import { markStyles, pdfPageStyles } from 'styles/reader/Document/PDF';
import { commentIcon } from 'app/components/RenderFunctions';
import { showPage } from 'store/reader/documentViewer';

/**
 * PDF Page Component
 * @param {Object} props -- Contains the PDF Page props and functions to manipulate the rotation/scale/dimensions
 */
export const Page = ({
  addingComment,
  scale,
  pageRef,
  canvasRef,
  currentDocument,
  rotation,
  style,
  numColumns,
  rowIndex,
  columnIndex,
  moveComment,
  match,
  moveMouse,
  getCoords,
  currentPageIndex,
  ...props
}) => {
  // Calculate the Page Index
  const pageIndex = (numColumns * rowIndex) + columnIndex;

  useEffect(() => {
    if (currentDocument?.id) {
      showPage({
        pageIndex,
        scale,
        rotation: currentDocument.rotation,
        docId: match.params.docId,
        currentPage: pageIndex + 1,
      });

    }
  }, [currentDocument?.id, scale, currentDocument?.rotation]);

  return pageIndex >= currentDocument.numPages ? (
    <div key={(numColumns * rowIndex) + columnIndex} style={style} />
  ) : (
    <div key={pageIndex} style={style} >
      <div
        id={`pageContainer${pageNumber(pageIndex)}`}
        className={classNames({
          page: true,
          'cf-pdf-pdfjs-container': true,
          'cf-pdf-placing-comment': addingComment
        })}
        ref={pageRef}
        {...markStyles}
      >
        <div id={`rotationDiv${pageNumber(pageIndex)}`} style={pdfPageStyles(rotation, style.height, style.width)}>
          <canvas id={`pdf-canvas-${currentDocument.id}-${pageIndex}`} ref={canvasRef} className="canvasWrapper" />
          {addingComment && (
            <div
              id={`canvas-cursor-${pageIndex}`}
              className="commentIcon-container canvas-cursor"
              style={{
                position: 'absolute',
                display: pageIndex === currentPageIndex ? 'block' : 'none',
                left: '50%',
                top: '50%'
              }}
            >
              {commentIcon()}
            </div>
          )}
          <div className="cf-pdf-annotationLayer">
            <Comments
              {...props}
              getCoords={getCoords}
              handleDrop={(event) => moveComment(getCoords(event, pageIndex), pageIndex)}
              moveMouse={(event) => moveMouse(getCoords(event, pageIndex), pageIndex)}
              currentDocument={currentDocument}
              documentId={currentDocument.id}
              pageIndex={pageIndex}
              scale={scale}
              dimensions={{ width: style.width, height: style.height }}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

Page.propTypes = {
  currentPageIndex: PropTypes.number,
  pageIndex: PropTypes.number,
  addingComment: PropTypes.bool,
  isVisible: PropTypes.bool,
  getCoords: PropTypes.func,
  moveComment: PropTypes.func,
  scale: PropTypes.number,
  pageRef: PropTypes.element,
  canvasRef: PropTypes.element,
  currentDocument: PropTypes.object,
  rotation: PropTypes.number,
  style: PropTypes.object,
  match: PropTypes.object,
  numPages: PropTypes.number,
  numColumns: PropTypes.number,
  rowIndex: PropTypes.number,
  columnIndex: PropTypes.number,
  setPageNumber: PropTypes.func,
  moveMouse: PropTypes.func,
};
