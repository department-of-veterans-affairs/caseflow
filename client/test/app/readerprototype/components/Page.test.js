import { render } from '@testing-library/react';
import Page from 'app/readerprototype/components/Page';
import { def, get } from 'bdd-lazy-var/getter';
import fs from 'fs';
import { getDocument } from 'pdfjs-dist/es5/build/pdf';
import React from 'react';

window.IntersectionObserver = jest.fn(() => ({
  observe: jest.fn(),
  disconnect: jest.fn()
}));

def(
  'render',
  () => async () => {
    const pageData = async () => {
      const raw = fs.readFileSync('test/fixtures/pdfs/Informal_Form9.pdf');
      const arrayBuffer = raw.buffer;

      const pdf = await getDocument({ data: arrayBuffer }).promise;

      return await pdf.getPage(1);
    };

    return render(
      <Page page={await pageData()} scale={100} />
    );
  }
);

it('renders', async () => {
  const { container } = await get.render();

  expect(container).toMatchSnapshot();
});
