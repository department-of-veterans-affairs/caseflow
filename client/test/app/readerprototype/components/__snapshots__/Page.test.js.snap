// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div
    class="prototype-canvas-wrapper"
    id="canvasWrapper-1"
    style="height: 792px; width: 612px; position: relative; background-color: white;"
  >
    <canvas
      class="prototype-canvas "
      height="792"
      id="canvas-1"
      style="position: relative; top: 0px;"
      width="612"
    />
  </div>
</div>
`;
